cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/AvroFileManager.java

```diff
@@ -0,0 +1,300 @@
+    long currentTs = System.currentTimeMillis();
+    long timeSinceFileCreate = currentTs - avroFile.getCreateTime();
+    if (timeSinceFileCreate > maxLifetimeMillis) {
+      avroFile.close();
+    }
+    if (!avroFile.isOpen()) {
+      LOG.info("Rotating a closed file {}", avroFile.getLocation());
+      return createAvroFile(identifier, timestamp);
+    }
+    return avroFile;
+  }
+
+  /**
+   * closes all the open avro files in the map
+   */
+  public void close() {
+    // close all the files in fileMap
+    // clear the map
+    for (AvroFile file : fileMap.values()) {
+      try {
+        file.flush();
+        file.close();
+      } catch (Exception e) {
+        LOG.debug("Exception while flushing and closing contents to file {}", file.getLocation(), e);
+      }
+    }
+    fileMap.clear();
+  }
+
+  private void closeAndDelete(AvroFile avroFile) {
+    try {
+      try {
+        avroFile.close();
+      } finally {
+        if (avroFile.getLocation().exists()) {
+          avroFile.getLocation().delete();
+        }
+      }
+    } catch (IOException e) {
+      LOG.error("Error while closing and deleting file {}", avroFile.getLocation(), e);
+    }
+  }
+
+  /**
+   * flushes the contents of all the open avro files
+   * @throws IOException
+   */
+  public void flush() throws IOException {
+    // perform flush on all the files in the fileMap
+    for (AvroFile file : fileMap.values()) {
+      try {
+        file.flush();
+      } catch (Exception e) {
+        LOG.debug("Exception while flushing contents to file {}", file.getLocation());
+        throw Throwables.propagate(e);
+      }
+    }
+  }
+
+  /**
+   * Represents an Avro file.
+   *
+   * Since there is no way to check the state of the underlying file on an exception,
+   * all methods of this class assume that the file state is bad on any exception and close the file.
+   */
+  public class AvroFile implements Closeable {
+    private final Location location;
+    private FSDataOutputStream outputStream;
+    private DataFileWriter<GenericRecord> dataFileWriter;
+    private long createTime;
+    private boolean isOpen = false;
+
+    public AvroFile(Location location) {
+      this.location = location;
+    }
+
+    /**
+     * Opens the underlying file for writing.
+     * If open throws an exception then underlying file may still need to be deleted.
+     *
+     * @throws IOException
+     */
+    void open() throws IOException {
+      try {
+        this.outputStream = new FSDataOutputStream(location.getOutputStream(), null);
+        this.dataFileWriter = new DataFileWriter<>(new GenericDatumWriter<GenericRecord>(schema));
+        this.dataFileWriter.create(schema, this.outputStream);
+        this.dataFileWriter.setSyncInterval(syncIntervalBytes);
+        this.createTime = System.currentTimeMillis();
+        // Sync the file as soon as it is created, otherwise a zero length Avro file can get created on OOM
```

Comment:
though the CDAP Log appender wouldn't go OOM, log processors could cause the log saver to go OOM right ? should we keep this or remove ?